receiver/postgresqlreceiver/scraper.go

```diff
@@ -0,0 +1,344 @@
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.opentelemetry.io/collector/receiver/scrapererror"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/postgresqlreceiver/internal/metadata"
+)
+
+type postgreSQLScraper struct {
+	logger        *zap.Logger
+	config        *Config
+	clientFactory postgreSQLClientFactory
+}
+
+type postgreSQLClientFactory interface {
+	getClient(c *Config, database string) (client, error)
+}
+
+type defaultClientFactory struct{}
+
+func (d *defaultClientFactory) getClient(c *Config, database string) (client, error) {
+	return newPostgreSQLClient(postgreSQLConfig{
+		username: c.Username,
+		password: c.Password,
+		database: database,
+		tls:      c.TLSClientSetting,
+		address:  c.NetAddr,
+	})
+}
+
+func newPostgreSQLScraper(
+	logger *zap.Logger,
+	config *Config,
+	clientFactory postgreSQLClientFactory,
+) *postgreSQLScraper {
+	return &postgreSQLScraper{
+		logger:        logger,
+		config:        config,
+		clientFactory: clientFactory,
+	}
+}
+
+// initMetric initializes a metric with a metadata attribute.
+func initMetric(ms pdata.MetricSlice, mi metadata.MetricIntf) pdata.Metric {
+	m := ms.AppendEmpty()
+	mi.Init(m)
+	return m
+}
+
+// addToIntMetric adds and attributes a int sum datapoint to metricslice.
+func addToIntMetric(metric pdata.NumberDataPointSlice, attributes pdata.AttributeMap, value int64, ts pdata.Timestamp) {
+	dataPoint := metric.AppendEmpty()
+	dataPoint.SetTimestamp(ts)
+	dataPoint.SetIntVal(value)
+	if attributes.Len() > 0 {
+		attributes.CopyTo(dataPoint.Attributes())
+	}
+}
+
+// scrape scrapes the metric stats, transforms them and attributes them into a metric slices.
+func (p *postgreSQLScraper) scrape(ctx context.Context) (pdata.Metrics, error) {
+	// metric initialization
+	md := pdata.NewMetrics()
+	ilm := md.ResourceMetrics().AppendEmpty().InstrumentationLibraryMetrics().AppendEmpty()
+	ilm.InstrumentationLibrary().SetName("otel/postgresql")
+	now := pdata.NewTimestampFromTime(time.Now())
+
+	blocksRead := initMetric(ilm.Metrics(), metadata.M.PostgresqlBlocksRead).Sum().DataPoints()
+	commits := initMetric(ilm.Metrics(), metadata.M.PostgresqlCommits).Sum().DataPoints()
+	databaseSize := initMetric(ilm.Metrics(), metadata.M.PostgresqlDbSize).Gauge().DataPoints()
+	backends := initMetric(ilm.Metrics(), metadata.M.PostgresqlBackends).Gauge().DataPoints()
+	databaseRows := initMetric(ilm.Metrics(), metadata.M.PostgresqlRows).Gauge().DataPoints()
+	operations := initMetric(ilm.Metrics(), metadata.M.PostgresqlOperations).Sum().DataPoints()
+	rollbacks := initMetric(ilm.Metrics(), metadata.M.PostgresqlRollbacks).Sum().DataPoints()
+
+	var errors scrapererror.ScrapeErrors
+
+	databases := p.config.Databases
+	listClient, err := p.clientFactory.getClient(p.config, "")
+	if err != nil {
+		p.logger.Error("Failed to initialize connection to postgres", zap.Error(err))
+		errors.Add(err)
```

Comment:
Is it certain that this has been instantiated?

Reply:
From what I can tell there is no additional required instantiation.